src/compiler/checker.ts

```diff
@@ -26638,12 +26638,17 @@ namespace ts {
                 if (contextualReturnType) {
                     const functionFlags = getFunctionFlags(func);
                     if (functionFlags & FunctionFlags.Generator) { // Generator or AsyncGenerator function
-                        const use = functionFlags & FunctionFlags.Async ? IterationUse.AsyncGeneratorReturnType : IterationUse.GeneratorReturnType;
-                        const iterationTypes = getIterationTypesOfIterable(contextualReturnType, use, /*errorNode*/ undefined);
-                        if (!iterationTypes) {
+                        if (contextualReturnType.flags & TypeFlags.Union) {
+                            contextualReturnType = filterType(contextualReturnType, type => {
+                                const generator = createGeneratorReturnType(anyType, anyType, anyType, (functionFlags & FunctionFlags.Async) !== 0);
+                                return checkTypeAssignableTo(generator, type, /*errorNode*/ undefined);

```

Comment:
`getIterationTypeOfGeneratorFunctionReturnType` will do this work for you (returning `undefined` if its not generator-like), would this work instead?

```ts
const isAsyncGenerator = (functionFlags & functionFlags.Async) !== 0;
const iterationReturnType = mapType(contextualReturnType,
    type => getIterationTypeOfGeneratorFunctionReturnType(IterationTypeKind.Return, type, isAsyncGenerator));
if (iterationReturnType) {
    return undefined;
}
```

`mapType` will remove `undefined` results, and will return `undefined` if all results were `undefined`, so it can also act as a filter.

Reply:
The problem is that `getIterationTypeOfGeneratorFunctionReturnType` returns, e.g., `any` for type `number`, which is why I went for removing the useless union components in the first place. Should I somehow change the behavior of `getIterationTypeOfGeneratorFunctionReturnType` instead? That seemed more tricky.

Reply:
It returns `any` if the type passed is a number type? It's only supposed to return `any` if the type passed in is `any`. For anything else that isn't iterable it's supposed to return `undefined`.

Reply:
Yes, it returns any. What happens is: we call `getIterationTypeOfGeneratorFunctionReturnType` with type `number`. This calls `getIterationTypesOfGeneratorFunctionReturnType`, which then returns `getIterationTypesOfIterable(number, use, /*errorNode*/ undefined) || getIterationTypesOfIterator(number, resolver, /*errorNode*/ undefined)`.

`getIterationTypesOfIterable(number, use, /*errorNode*/ undefined)` returns `undefined`, so we return `getIterationTypesOfIterator(number, resolver, /*errorNode*/ undefined)`.

`getIterationTypesOfIterator(number, resolver, /*errorNode*/ undefined)` eventually just tries to get the types of properties `next`, `return`, `throw`, which are all `any` for type `number`. So it returns something like `IteratorType { next = any, throw = any, return = any }`, from which we conclude that the generator return type of type `number` is `any`.
So yeah, in short, `getIterationTypeOfGeneratorFunctionReturnType` returns `any` for a non-iterable/iterator type like `number`, it doesn't return `undefined`.

I don't understand all the subtleties involved, so that's why I went with filtering the union by checking generator type assignability, but maybe I could be filtering using some of the other `getIterationTypesX` functions?

I think filtering by calling `getIterationTypesOfIterable` and seeing if it's undefined could maybe work at first, but the fact that when it returns `undefined` we also call `getIterationTypesOfIterator` seemed to suggest to me that `getIterationTypesOfIterable` didn't cover all of the cases of iterator/iterable types.

Let me know what you think.

Reply:
It seems `getIterationTypesOfIteratorSlow` may be doing the wrong thing here. It should be returning `noIterationTypes` for `number`:

```ts
function getIterationTypesOfIteratorSlow(type: Type, resolver: IterationTypesResolver, errorNode: Node | undefined) {
    const iterationTypes = combineIterationTypes([
        getIterationTypesOfMethod(type, resolver, "next", errorNode),
        getIterationTypesOfMethod(type, resolver, "return", errorNode),
        getIterationTypesOfMethod(type, resolver, "throw", errorNode),
    ]);
    return setCachedIterationTypes(type, resolver.iteratorCacheKey, iterationTypes);
}
```

- `getIterationTypesOfMethod(number, resolver, "next", errorNode)` - This is probably the culprit, as it should return `undefined` for type `number`, but instead returns `anyIterationTypes` due to this logic:
    ```ts
    // Both async and non-async iterators *must* have a `next` method.
    const methodSignatures = methodType ? getSignaturesOfType(methodType, SignatureKind.Call) : emptyArray;
    if (methodSignatures.length === 0) {
        if (errorNode) {
            const diagnostic = methodName === "next"
                ? resolver.mustHaveANextMethodDiagnostic
                : resolver.mustBeAMethodDiagnostic;
            error(errorNode, diagnostic, methodName);
        }
        return methodName === "next" ? anyIterationTypes : undefined;
     // ^- should probably just be `return undefined;`
    }
    ```
- `getIterationTypesOfMethod(number, resolver, "return", errorNode)` - This looks like it returns `undefined` for type `number`, which is correct.
- `getIterationTypesOfMethod(number, resolver, "throw", errorNode)` - This looks like it returns `undefined` for type `number`, which is correct.

The results of those three methods are passed to `combineIterationTypes`, which would then return `noIterationTypes` once the iteration types for the `"next"` method are corrected.

Reply:
I need to look into this a bit further, as changing this results in some other additional diagnostics that we may not want.

Reply:
The return value for `getIterationTypesOfIteratorSlow` should be fixed by #50146.

Reply:
I ended up not using `mapType` in the way you suggested, because in the previous version of the code, when we call `getIterationTypeOfGeneratorFunctionReturnType` on a union, although it maps the workers to each component type in the union, the way it combines the resulting iterationTypes is specific, so I don't think we'd get the same result if we just called `mapType`, which just unions the resulting mapped types.

Reply:
I think it should have produced the same type for `IterationTypeKind.Return`, since the logic in `getIterationTypesOfIterable` related to union types just appends the resulting types and unions them for `yield` and `return`, with a bit more in the way of caching and error reporting. The only case that would differ would be for `IterationTypeKind.Next`, since the `next` types are intersected rather than unioned.